proposals/0008-c-interface-compiler-library.md

```diff
@@ -996,21 +998,535 @@ if (compileResult->HasOutput(DXC_OUT_OBJECT)) {
+* 
+* \param timeout the amount of time in milliseconds to wait for the
+* operation. CompilerLibraryError_WaitTimeout will be returned if the
+* timeout is hit before the operation has completed.
+*/
+CompilerLibraryErrorCode clang_WaitForCompilation(
+    CompilerLibraryInstance instance,
+    CompilerLibraryCookie cookie,
+    uint32_t timeout);
+
+/**
+* Callback to retrieve a buffer for the specified include path.
+*
+* This callback is implemented and passed to compile for 
+* functions for supplying custom buffers for the specified
+* include path during compilation.
+*
+* \param path include path
+*
+* \param buffer the buffer with contents for the specified include path. This
+* buffer is owned by the caller.
+*
+* \param bufferSize [out] The size of the buffer in bytes
+*
+* \returns a bool indicating if the contents of the buffer have been filled
+* successfully.
+*/
+typedef bool (*CompilerIncludeCallback) (
+    const wchar_t* path,
+    char** buffer,
+    size_t* bufferSize);
+```
+
+### Include handlers
+Include handler support works very similar to the DXC compiler library except
+the handler is now a callback function instead of an interface.  The callback
+function is passed in as an optional parameter to the compile entrypoints.
+
+```c++
+/**
+* Callback to retrieve a buffer for the specified include path.
+*
+* This callback is implemented and passed to compile for 
+* functions for supplying custom buffers for the specified
+* include path during compilation.
+*
+* \param path include path
+*
+* \param buffer the buffer with contents for the specified include path. This
+* buffer is owned by the caller.
+*
+* \param bufferSize [out] The size of the buffer in bytes
+*
+* \returns a bool indicating if the contents of the buffer have been filled
+* successfully.
+*/
+typedef bool (*CompilerIncludeCallback) (
+    const wchar_t* path,
+    char** buffer,
+    size_t* bufferSize);
 
-This approach combines Alternative 1 and adds compiling the wrapper code into
-formal header/libraries that can be consumed by legacy toolchains.
+```
+
+### clang Library Api code snippets
+The following snippet shows how a shader is compiled using the library.
+NOTE: RAII wrappers could be created and used to auto-free returned objects,
+but they are not used here to show raw usage of the api.
+
+#### Synchronous Compilation of a shader
+```c++
+inline std::vector<char> ReadEntireFile(char const* filename)
+{
+  std::ifstream fs(filename, std::ios_base::binary | std::ios_base::in);
+  if (!fs.is_open())
+      throw std::runtime_error("Failed to open file");
+
+  fs.seekg(0, std::ios_base::end);
+  auto length = fs.tellg();
+  fs.seekg(0, std::ios_base::beg);
+
+  std::vector<char> data;
+  data.resize(static_cast<size_t>(length));
+
+  fs.read(reinterpret_cast<char*>(data.data()), length);
+  return data;
+}
+
+int main()
+{
+  CompilerLibraryInstance compiler = clang_createCompilerLibrary();
+  if (compiler)
```

Comment:
Hmm.. Other apis use the errno error system for this type of thing, but that isn't in this design as the error result mechanism.  Guess I could make the errorcode be an in/out param to the api and it could be communicated from there?